s3/src/main/scala/akka/stream/alpakka/s3/impl/S3Stream.scala

```diff
@@ -131,13 +186,23 @@ import scala.util.{Failure, Success, Try}
       }
   }.mapMaterializedValue(_ => NotUsed)
 
-  sealed trait ListBucketState
+  /**
+   * An ADT that represents the current state of pagination
+   */
+  sealed trait S3PaginationState[T]
 
-  case object Starting extends ListBucketState
+  final case class Starting[T]() extends S3PaginationState[T]
```

Comment:
Ideally both `Starting` and `Finished` should be case objects however I didn't manage to get type inference to work nicely so I ended up converting them to a case class that has a zero parameters since thats the only way to propagate the `T` across. 

Maybe there is a nicer solution for this?

Reply:
Let's not fight for it. It is internal API anyway.